socs/util.py

```diff
@@ -1,9 +1,195 @@
+class BaseAction:
+    """
+    Base subclass for actions that correspond to OCS tasks. Such actions can
+    be used to generate a generic task that creates the action, passes
+    to some callback function, and waits until the action is resolved before
+    returning.
+    """
+
+    def __post_init__(self) -> None:
+        self._session_data: Dict[str, Any] = {}
+        self._processed: bool = False
+        self._success: bool = False
+        self._traceback: Optional[str] = None
+        self._return_message: Optional[str] = None
+
+    def resolve_action(
+        self,
+        success: bool,
+        traceback: Optional[str] = None,
+        return_message: Optional[str] = None,
+    ) -> None:
+        """
+        Resolves an action, signifying it has been completed. Tasks waiting for
+        this action to be resolved can then return.
+        """
+        self._success = success
+        if traceback is not None:
+            self._traceback = traceback
+        if return_message is not None:
+            self._return_message = return_message
+        self._processed = True
+
+    def update_session_data(self, data: Dict[str, Any]):
+        self._session_data.update(data)
+
+    def sleep_until_resolved(self, session: ocs_agent.OpSession, interval=0.2) -> None:
+        """
+        Sleeps until the action has been resolved.
+        """
+        while not self._processed:
+            session.data = self._session_data
+            time.sleep(interval)
+
+        session.data = self._session_data
+
+
+def is_instanceable(t: Type) -> bool:
+    """
+    Checks if its possible to run isinstance with a specified type. This is
+    needed because older version of python don't let you run this on subscripted
+    generics.
+    """
+    try:
+        isinstance(0, t)
+        return True
+    except Exception:
+        return False
+
+
+def get_param_type(t: Type) -> Optional[Type]:
+    """
+    OCS param type variables require you to be able to run isinstance,
+    which does not work for subscripted generics like Optional in python3.8.
+    This function attempts to convert types to values that will be accepted
+    by ocs_agent.param, unwrapping optional types if we are unable to run
+    isinstance off the bat.
+
+    Other subscripted generics such as List[...] or Dict[...] are not currently
+    supported.
+
+    This function will return the unwrapped type, or None if it fails.
+    """
+    origin_type = get_origin(t)
+
+    # Unwrap possible option type
+    if origin_type == Union:
+        sub_types = get_args(t)
+        if len(sub_types) != 2:
+            return None
+        if type(None) not in sub_types:
+            return None
+        for st in sub_types:
+            if st is not type(None):
+                if is_instanceable(st):
+                    return st
+
+    elif is_instanceable(t):
+        # If this works, then it should work with ocs_agent.param
+        return t
+
+    return None
```

Comment:
Let's mark these two functions as private with a leading `_`. They can always be made public later if needed.